### PR TITLE
Handle scenarios for requests/responses where rawBody is being used

### PR DESCRIPTION
**Pull request by kkoomen**
Hi,

We're having a project where we use the Throttler being defined globally in the app module and we're listening for stripe webhook events with the [@golevelup/nestjs-stripe](https://www.npmjs.com/package/@golevelup/nestjs-stripe) module.

We've added `{ bodyParser: false }` inside the `main.ts` followed by defining the following setup in the app module:

```ts
export class AppModule implements NestModule {
  configure(consumer: MiddlewareConsumer): void {
    consumer
      .apply(RawBodyMiddleware)
      .forRoutes({
        path: 'stripe/webhook',
        method: RequestMethod.ALL,
      })
      .apply(JsonBodyMiddleware)
      .forRoutes('*')
  }
}
```

After triggering a Stripe event, this resulted in the following error:

```
TypeError: Cannot read property 'header' of undefined
    at ThrottlerGuard.handleRequest (/Users/<username>/projects/<my-project>/node_modules/@nestjs/throttler/dist/throttler.guard.js:68:13)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async GuardsConsumer.tryActivate (/Users/<username>/projects/<my-project>/node_modules/@nestjs/core/guards/guards-consumer.js:16:17)
    at async canActivateFn (/Users/<username>/projects/<my-project>/node_modules/@nestjs/core/helpers/external-context-creator.js:162:33)
    at async target (/Users/<username>/projects/<my-project>/node_modules/@nestjs/core/helpers/external-context-creator.js:75:31)
    at async Object.handler (/Users/<username>/projects/<my-project>/node_modules/@nestjs/core/helpers/external-proxy.js:9:24)
    at async Promise.all (index 0)
    at async StripeWebhookService.handleWebhook (/Users/<username>/projects/<my-project>/node_modules/@golevelup/nestjs-stripe/lib/stripe.module.js:108:17)
    at async StripeWebhookController.handleWebhook (/Users/<username>/projects/<my-project>/node_modules/@golevelup/nestjs-stripe/lib/stripe.webhook.controller.js:34:9)
```

Basically, the problem is that in this scenario, the `res` is `undefined` in the throttler guard and I don't really know why, but adding an if-statement around it will fix the problem.

I thought to myself, I could add `ignoreUserAgents: [/^Stripe\/[\d.]+/]` as well, but this resulted in the error `Cannot read property 'user-agent' of undefined` so I've also improved that check by checking if `req.headers` exists.

There's no way for me to use stripe with throttler globally configured together, so we're disabling throttler in our project at this point. I suggest we merge this crucial change ASAP and do a deploy to NPM as well.

### CODE REVIEW

**Comment by kkoomen** [comment]
@jmcdo29 Can you help me to fix the tests? It seems to fail quite a lot.

**Comment by jmcdo29** [comment]
@kkoomen we can probably remove the tests for node 10. 12 and 14 are current anyways. 

Should we make a check in here for `if(!res) { return true; }` so we don't have to wrap the entire thing?

Also, isn't this a perfect use case for the `@ThrottlerSkip()` decorator?

**Comment by kkoomen** [comment]
> Should we make a check in here for `if(!res) { return true; }` so we don't have to wrap the entire thing?

I've though about it, but then it wouldn't take this endpoint in consideration for the throttling. I still want this endpoint to be throttled though so that's why I did add `if (res) { .... }` around the parts that contain `res.header()`.

**Comment by jmcdo29** [comment]
I'm going to end up opening a PR in the `@golevelup` package to allow for decorators on the endpoint that would allow for calling `@ThrottlerSkip()`. I'll keep reviewing this in the meantime though

**Comment by ZhaoKunLong** [comment]
@kkoomen Bro. How's it going, nice news! I checked it in the @goleveup/nestjs-stripe @jmcdo29  his  [PR262](https://github.com/golevelup/nestjs/pull/262) is merged. so we can use it like this now.
```js
StripeBaseModule.forRootAsync(StripeBaseModule, {
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: (configService: ConfigService) => ({
        apiKey: configService.get<string>('STRIPE_SECRET_KEY'),
        timeout: 1000 * 20,
        webhookConfig: {
          stripeWebhookSecret: configService.get<string>('STRIPE_WEBHOOK_SECRET_KEY'),
          decorators: [SkipThrottle()],
        },
      }),
    }),
```
And I will remove the  `ignoreUserAgents` is right?

```js
ThrottlerModule.forRootAsync({
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: (configService: ConfigService) => ({
        ttl: 60,
        limit: 1000,
//        ignoreUserAgents: [/^Stripe\/[\d.]+/],
        storage: new ThrottlerStorageRedisService(configService.get<string>('REDIS_URL')),
      }),
    }),
```

**Comment by kkoomen** [comment]
@ZhaoKunLong Yes, you can remove the `ignoreUserAgents` key from the `ThrottlerModule` and add `decorators: [SkipThrottle()]` in the StripeBaseModule defined in the `app.module.ts`.

@jmcdo29 Seems like this issue is solved on one side, but since I did add additional checks, shall we merge those or are those not needed anymore?

**Comment by jmcdo29** [comment]
I think for now we can close this. I'm not a huge fan of just wrapping this all in an `if` in the case that `res` is undefined, which is why I wanted to allow the fix at the StripeModule level. If this comes up again though we can revisit the PR
